### DISCUSSION TITLE
Freebsd

### DISCUSSION THREAD

**Discussion by Mark0x01** [modoboa/modoboa#3142] [Installer]
Is there any interest in making an official FreeBSD capable installer ?

I see a few users reporting issues that run Modoboa on FreeBSD.

I have done some work on enhancing the installer to support FreeBSD, and it was about 80% working around version 2.0.0 and used that as my base install.

**Comment by tonioo** [comment]
Supporting FreeBSD would be great so we're interested by any PR ;)

**Reply by Mark0x01** [reply]
Ok, for the brave who follow, for FreeBSD about a dozen packages need to be built from ports, as the required options aren't there in the standard prebuilt package, which makes it a long process. Python packages are best installed using pip, for the same reason.

I'll complete a working version of the installer first with freebsd support added and post a link to it if anyone else wants to test it with build instructions for the port builds, which need to be installed first.

This will take a while, as I'm still finding issues that require packages to be rebuilt, as I didn't add everything at the begining.

Latest tricky one I'm working on is adding automx support to my production build. When I add automx to uwsgi, it breaks the uwsgi, taking out the web interfaces. Looks to be missing a needed build option.

**Comment by Mark0x01** [comment]
Getting there, but merging changes since my initial working version about 12 months ago keeps breaking things that may be FreeBSD specific.
e.g. latest updated utils.exec_cmd()  now hangs, and the added python.get_package_version() fails to work as expected.

Not sure why, but pip doesn't seem to bring in the most recent modoboa-postfix-autoreply, causing further unexpected issues.

**Comment by Mark0x01** [comment]
This has come to a halt, due to use of psycopg[binary] in 2.6.5, which is unsupported on BSD. 
Although a psycopg[c] would work, since modoboa is installed from PyPy, I don't see a way around the issue.

**Reply by kryskool** [reply]
With @tonioo we have update the python dependancies to use psycopg[c] with FreeBSD otherwise we use psycopg[binary]

Can you confirm it works ?

**Comment by Mark0x01** [comment]
Progress report:

This is proving quite challenging. 

For Modoboa 2.7.0, I have a working installer for FreeBSD 13.5, which didn't work on 14.3, but I'm making slow but steady progress to make it reliable and more easily modified in the future.

One nasty and unexpected problem is disrupting my dev platform: Very intermittently failing zfs cache and log ssd's in my Proxmox host used for testing.
When they drop out, both togeather and I suspect it is a sas controller issue, a package installs but with missing files, so breaks.
There might also be other corruption issues, so I'll need to spend some time fixing it first.

**Comment by Mark0x01** [comment]
The first release of a FreeBSD compatable installer is now available [here](https://github.com/Mark0x01/modoboa-installer).

Tested with FreeBSD 14.3, python 3.11, postgres and amavis.

**Reply by tonioo** [reply]
Do you plan to create a PR soon?

**Reply by Mark0x01** [reply]
The changes may be too extensive ?

I still have a few gremlins to sort, and then I need to test further to ensure I haven't broken the linux side.

Would prefer to have had a few testers try it to help weed out issues I haven't spotted and get some feedback.
